### ISSUE TITLE
Potential data race detected by TSAN

### ISSUE THREAD

**Issue by kluever** [ben-manes/caffeine#556]
One of our tools (TSAN) detected a potential data race in Caffeine during a unit test. I've included as much of the stack trace as I can below. Hopefully it can point you in the right direction?

```java
WARNING: ThreadSanitizer: data race (pid=9172)
  Write of size 4 at 0x0000cf0c61d0 by thread T33 (mutexes: write M110619669326012648, write M861313432178006392):
    #0 com.github.benmanes.caffeine.cache.BaseMpscLinkedArrayQueue.resize(J[Ljava/lang/Object;JLjava/lang/Object;)V MpscGrowableArrayQueue.java:533 
    #1 com.github.benmanes.caffeine.cache.BaseMpscLinkedArrayQueue.offer(Ljava/lang/Object;)Z MpscGrowableArrayQueue.java:250 
    #2 com.github.benmanes.caffeine.cache.BoundedLocalCache.afterWrite(Ljava/lang/Runnable;)V BoundedLocalCache.java:1356 
    #3 com.github.benmanes.caffeine.cache.BoundedLocalCache.replace(Ljava/lang/Object;Ljava/lang/Object;Ljava/lang/Object;)Z BoundedLocalCache.java:2271 
    #4 com.github.benmanes.caffeine.cache.LocalAsyncCache.lambda$handleCompletion$7(Ljava/util/concurrent/atomic/AtomicBoolean;JLjava/lang/Object;Ljava/util/concurrent/CompletableFuture;ZLjava/lang/Object;Ljava/lang/Throwable;)V LocalAsyncCache.java:207 
    #5 com.github.benmanes.caffeine.cache.LocalAsyncCache$$Lambda$329.accept(Ljava/lang/Object;Ljava/lang/Object;)V ?? 
    #6 java.util.concurrent.CompletableFuture.uniWhenComplete(Ljava/lang/Object;Ljava/util/function/BiConsumer;Ljava/util/concurrent/CompletableFuture$UniWhenComplete;)Z CompletableFuture.java:859 
    #7 java.util.concurrent.CompletableFuture.uniWhenCompleteStage(Ljava/util/concurrent/Executor;Ljava/util/function/BiConsumer;)Ljava/util/concurrent/CompletableFuture; CompletableFuture.java:883 
    #8 java.util.concurrent.CompletableFuture.whenComplete(Ljava/util/function/BiConsumer;)Ljava/util/concurrent/CompletableFuture; CompletableFuture.java:2251 
    #9 com.github.benmanes.caffeine.cache.LocalAsyncCache.handleCompletion(Ljava/lang/Object;Ljava/util/concurrent/CompletableFuture;JZ)V LocalAsyncCache.java:190 
    #10 com.github.benmanes.caffeine.cache.LocalAsyncCache.put(Ljava/lang/Object;Ljava/util/concurrent/CompletableFuture;)V LocalAsyncCache.java:183 

```

**Comment by kluever** [comment]
```java
  Previous read of size 4 at 0x0000cf0c61d0 by thread T32 (mutexes: write M416301493996909720, write M416864443950331400, write M416301494012652224, write M416864443966073624):
    #0 com.github.benmanes.caffeine.cache.BaseMpscLinkedArrayQueue.offer(Ljava/lang/Object;)Z MpscGrowableArrayQueue.java:236 
    #1 com.github.benmanes.caffeine.cache.BoundedLocalCache.afterWrite(Ljava/lang/Runnable;)V BoundedLocalCache.java:1356 
    #2 com.github.benmanes.caffeine.cache.BoundedLocalCache.put(Ljava/lang/Object;Ljava/lang/Object;Lcom/github/benmanes/caffeine/cache/Expiry;ZZ)Ljava/lang/Object; BoundedLocalCache.java:2008 
    #3 com.github.benmanes.caffeine.cache.BoundedLocalCache.put(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object; BoundedLocalCache.java:1956 
    #4 com.github.benmanes.caffeine.cache.LocalAsyncCache.put(Ljava/lang/Object;Ljava/util/concurrent/CompletableFuture;)V LocalAsyncCache.java:182 
```

**Comment by ben-manes** [comment]
Thanks @kluever. `MpscGrowableArrayQueue` is an embedded copy of JCTools'. This was changed to use VarHandles instead of `sun.misc.Unsafe` and to copy the type hierarchy into a single file instead of many (as shared by other queues). However, because that resulted in some (desirably equivalent) changes, it wouldn't be fair for us to expect @nitsanw's help just yet. cc'd only for an fyi.

Can you provide more details so that I can reproduce? If I can get it to fail, and especially if I can also make it fail using a direct jctools dependency, then it will help us narrow to the cause. I'd need to know the library version, TSAN setup, etc. as it is not part of my CI flow.

On a positive note, I have not heard of issues in this for a few years and this could be a benign race. I know that piggybacking on memory barriers is a trick that jsr166 code uses for efficiency, so TSAN may have observed a false positive.

**Comment by ben-manes** [comment]
It looks like you are using v2.8.5 or earlier, as the next release updates to JDK15 compatible padding. The TSAN error is regarding a plain read of the `producerBuffer`. The two code sections are,

https://github.com/ben-manes/caffeine/blob/59e30e908f4eb4b3dccea0dad15fdcc93e17984a/caffeine/src/main/java/com/github/benmanes/caffeine/cache/MpscGrowableArrayQueue.java#L529-L533

https://github.com/ben-manes/caffeine/blob/59e30e908f4eb4b3dccea0dad15fdcc93e17984a/caffeine/src/main/java/com/github/benmanes/caffeine/cache/MpscGrowableArrayQueue.java#L234-L237

---

This is a benign race which relies on piggybacking on a prior read. The case within the `resize()` method has already acquired a exclusive access via the producerIndex.
 
https://github.com/ben-manes/caffeine/blob/59e30e908f4eb4b3dccea0dad15fdcc93e17984a/caffeine/src/main/java/com/github/benmanes/caffeine/cache/MpscGrowableArrayQueue.java#L282-L288

That means that its read of the `producerBuffer` is safe as the CAS provided a load barrier and no other writer may modify the memory location. It can then update the field to a new array and publish the value by a store barrier when releasing the producerIndex.

https://github.com/ben-manes/caffeine/blob/59e30e908f4eb4b3dccea0dad15fdcc93e17984a/caffeine/src/main/java/com/github/benmanes/caffeine/cache/MpscGrowableArrayQueue.java#L555-L556

The 2nd code snippet is within a retry loop inside of `offer()` and performs a load barrier each iteration. Because it fails during the race to increment the producer index, it retries and spin waits until the resize completes. This repeats until the new array is visible and the store succeeds.

https://github.com/ben-manes/caffeine/blob/59e30e908f4eb4b3dccea0dad15fdcc93e17984a/caffeine/src/main/java/com/github/benmanes/caffeine/cache/MpscGrowableArrayQueue.java#L225-L232

In summary, JCTools leverages piggybacking on memory barriers for visibility to reduce cache coherence overhead. We do similar in Caffeine too. This is harder to reason about and TSAN correctly pointed out that it could be a bug if unintentional. However when reviewing it from the perspective of memory barriers and the developer's intent, it is safe and effective.

Thanks for reporting and let me know if you have any questions. Thanks @nitsanw for the elegant code example of this technique.

**Comment by nitsanw** [comment]
Thanks for your kind words @ben-manes 🙂

**Comment by ben-manes** [comment]
I'd need more to go off of, so anything sent privately is fine.

The field(s) being modified are only done so under that object's synchronized block. A stale read is fine and writes are guarded by the instance lock. In this case it is writing to null out the value to assist GC (if objects span across generations), which a concurrent `getIfPresent` would report as a cache miss if it obtained the entry before it was evicted and read the value afterwards. From the information provided I do not see a fatal race.

**Comment by ben-manes** [comment]
Thanks for the logs. The section following immediately after your snippet is,

```
  Previous read of size 4 at 0x0000ccd79018 by thread T10 (mutexes: write M1, write M2, write M3, write M4, write M5, write M6):
    #0 (Generated Stub) <null> 
    #1 java.lang.invoke.VarHandleObjects$FieldInstanceReadOnly.get(Ljava/lang/invoke/VarHandleObjects$FieldInstanceReadOnly;Ljava/lang/Object;)Ljava/lang/Object; VarHandleObjects.java:62 
    #2 java.lang.invoke.VarHandleGuards.guard_L_L(Ljava/lang/invoke/VarHandle;Ljava/lang/Object;Ljava/lang/invoke/VarHandle$AccessDescriptor;)Ljava/lang/Object; VarHandleGuards.java:36 
    #3 com.github.benmanes.caffeine.cache.PS.getValue()Ljava/lang/Object; PS.java:79 
    #4 com.github.benmanes.caffeine.cache.BoundedLocalCache.getIfPresent(Ljava/lang/Object;Z)Ljava/lang/Object; BoundedLocalCache.java:2179 
    #5 com.github.benmanes.caffeine.cache.LocalAsyncCache.getIfPresent(Ljava/lang/Object;)Ljava/util/concurrent/CompletableFuture; LocalAsyncCache.java:70 
```

That's actually the case that I described, a `getIfPresent` racing with an eviction. The order does not matter because either it will observe the last value (present) or null (absent). Both are correct and the read piggybacks on the memory barrier from the ConcurrentHashMap.get(key) call, which performed a volatile read. Since we won't observe memory tearing this is safe, and in a concurrent setting either result is an acceptable answer. Please suppress this issue as benign.

Thanks!
